CONTRIBUTING.md

```diff
@@ -1,24 +1,38 @@
 # How to Contribute
 
-Contributions to libCEED are welcome.
+Contributions to libCEED are encouraged.
 
 Please use a pull request to the appropriate branch ('maint' for
 backward-compatible bug fixes for the last stable release, 'master' for
 new features and everything else).
 
-Please make your commits well-organized and atomic, using `git rebase
---interactive` as needed.  Open an issue or RFC pull request to discuss
-any significant changes before investing time.  Check that the tests
-pass using `make prove` or `make test`.  If adding a new feature, please
-add or extend a test so that it is tested.
+Please make your commits well-organized and
+[atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention),
+using `git rebase --interactive` as needed.  Check that tests
+(including "examples") pass using `make prove-all`.  If adding a new
+feature, please add or extend a test so that your new feature is
+tested.
+
+In typical development, every commit should compile, be covered by the
+test suite, and pass all tests.  This improves the efficiency of
+reviewing and facilitates use of
+[`git bisect`](https://git-scm.com/docs/git-bisect).
+
+Open an issue or RFC (request for comments) pull request to discuss
+any significant changes before investing time.  It is useful to create
+a WIP (work in progress) pull request for any long-running development
```

Comment:
I've become a bit uncomfortable with how 'long-running' some of our branches get. We have put a lot of work into rebasing big PRs lately. Perhaps some discussion about the 'typical' length (time) or size of a PR is due? Active-passive was clearly an exception that was going to cause lots of work, but I feel like some of our PRs are taking longer than they need and ultimately costing us more effort.

Reply:
This is a constant tension and I'm not sure how to word a recommendation. The alternative of having untested (usually buggy) development code in 'master' is not better IMO. Breaking up a large feature into identifiable sub-features that can be tested and merged is related to the discussion of atomicity.

Reply:
Right, I do not want development code in master. I am also not sure how to word it, but it was important to me that we discuss as a whole libCEED team because I think we are making extra work for ourselves sometimes.